internal/pkg/pkg.go

```diff
@@ -77,26 +92,177 @@ func New(path string) (*Pkg, error) {
+	}
+	defer f.Close()
+
+	d := yaml.NewDecoder(f)
+	d.KnownFields(true)
+	if err = d.Decode(kf); err != nil {
+		return kf, fmt.Errorf("unable to parse %s: %w", kptfilev1alpha2.KptFileName, err)
 	}
-	return p.pipeline
+	return kf, nil
+}
+
+// Pipeline returns the Pipeline section of the pkg's Kptfile.
+// if pipeline is not specified in a Kptfile, it returns Zero value of the pipeline.
+func (p *Pkg) Pipeline() (*kptfilev1alpha2.Pipeline, error) {
+	kf, err := p.Kptfile()
+	if err != nil {
+		return nil, err
+	}
+	pl := kf.Pipeline
+	if pl == nil {
+		return &kptfilev1alpha2.Pipeline{}, nil
+	}
+	return pl, nil
 }
 
 // String returns the slash-separated relative path to the package.
 func (p *Pkg) String() string {
 	return string(p.DisplayPath)
 }
+
+// SubPackages returns sub packages of a pkg.
+// A sub directory that has a Kptfile is considered a sub package.
+func (p *Pkg) SubPackages() (subPkgs []*Pkg, err error) {
+	pkgPath := string(p.UniquePath)
+
+	files, err := ioutil.ReadDir(pkgPath)
+	if err != nil {
+		return subPkgs, fmt.Errorf("failed to read sub dirs for %s %w", p, err)
+	}
+	for _, f := range files {
+		if f.IsDir() {
+			dirPath := path.Join(pkgPath, f.Name())
+			isKptPkg, err := kptfileutil.HasKptfile(dirPath)
+			if err != nil {
+				return subPkgs, fmt.Errorf("failed to check kptfile in subpackage %w", err)
+			}
+			if isKptPkg {
+				subPkg, err := New(dirPath)
+				if err != nil {
+					return subPkgs, fmt.Errorf("failed to read subpkg at path %s %w", dirPath, err)
+				}
+				subPkgs = append(subPkgs, subPkg)
+			}
+		}
+	}
+	sort.Slice(subPkgs, func(i, j int) bool {
+		return subPkgs[i].DisplayPath < subPkgs[j].DisplayPath
+	})
+	return subPkgs, nil
+}
+
+// LocalResources returns resources that belong to this package excluding the subpackage resources.
+func (p *Pkg) LocalResources(includeMetaResources bool) (resources []*yaml.RNode, err error) {
+	hasKptfile, err := kptfileutil.HasKptfile(p.UniquePath.String())
+	if err != nil {
+		return resources, fmt.Errorf("failed to check kptfile %w", err)
+	}
+	if !hasKptfile {
+		return nil, nil
+	}
+	pl, err := p.Pipeline()
+	if err != nil {
+		return nil, err
+	}
+
+	pkgReader := &kio.LocalPackageReader{
+		PackagePath:        string(p.UniquePath),
+		PackageFileName:    kptfilev1alpha2.KptFileName,
+		IncludeSubpackages: false,
+		MatchFilesGlob:     kio.MatchAll,
+	}
+	resources, err = pkgReader.Read()
+	if err != nil {
+		err = fmt.Errorf("failed to read resources for pkg %s %w", p, err)
+		return resources, err
+	}
+	if !includeMetaResources {
+		resources = filterMetaResources(resources)
+		resources, err = filterFunctionConfigFiles(resources, functionConfigFilePaths(pl))
+		if err != nil {
+			return resources, fmt.Errorf("failed to filter function config files: %w", err)
+		}
+	}
+	return resources, err
+}
+
+// filterMetaResources filters kpt metadata files such as Kptfile, function configs.
```

Comment:
I think we have defined meta resources as including function configs, no?

Reply:
Yes, done.